File: server/app/models/message.py
from sqlalchemy import Column, Integer, String, Text, DateTime
from sqlalchemy.orm import relationship
from datetime import datetime
from app.database import Base
from .base import LOCALTIME


MESSAGE_STATUS_STREAMING = 0
MESSAGE_STATUS_COMPLETED = 1

HAS_INTERACTIONS_PENDING = 0
HAS_INTERACTIONS_EXISTS = 1
HAS_INTERACTIONS_NONE = 2


class Message(Base):
    __tablename__ = "messages"

    id = Column(Integer, primary_key=True, index=True)
    session_id = Column(Integer, nullable=False, index=True)
    role = Column(String(20), nullable=False)
    content = Column(Text, nullable=False)
    status = Column(Integer, default=MESSAGE_STATUS_COMPLETED, nullable=False)
    has_interactions = Column(Integer, default=HAS_INTERACTIONS_NONE, nullable=False)
    created_at = Column(DateTime(timezone=True), default=datetime.now, server_default=LOCALTIME, nullable=False)
    updated_at = Column(DateTime(timezone=True), default=datetime.now, server_default=LOCALTIME, onupdate=datetime.now)

    session = relationship(
        "Session",
        back_populates="messages",
        primaryjoin="Message.session_id == Session.id",
        foreign_keys="Message.session_id"
    )

File: server/internal/model/message.go
// Package model defines the database models for the application.
package model

import "time"

// Message status constants.
const (
	MessageStatusStreaming  = 0 // Message is currently being generated (SSE streaming)
	MessageStatusCompleted = 1 // Message generation is complete

	// HasInteractions values indicate whether a message has associated agent interactions.
	HasInteractionsPending = 0 // Not yet determined (will be checked during processing)
	HasInteractionsExists  = 1 // Message has associated world-interaction records
	HasInteractionsNone    = 2 // Message has no world-interaction records
)

// Message represents a chat message in a session.
// Messages can be from the user (role="user") or the AI assistant (role="assistant").
// Assistant messages go through a streaming phase before being completed.
type Message struct {
	ID              int64     `gorm:"primaryKey;autoIncrement" json:"id"`
	SessionID       int64     `gorm:"not null;index;column:session_id" json:"session_id"`
	Role            string    `gorm:"type:varchar(20);not null" json:"role"`
	Content         string    `gorm:"type:text;not null" json:"content"`
	Status          int       `gorm:"not null;default:0" json:"status"`
	HasInteractions int       `gorm:"not null;default:0;column:has_interactions" json:"has_interactions"`
	CreatedAt       time.Time `gorm:"not null;autoCreateTime" json:"created_at"`
	UpdatedAt       time.Time `gorm:"not null;autoUpdateTime" json:"updated_at"`
}

func (Message) TableName() string { return "messages" }
